bakery/src/scripts/copy-resources-s3.py

```diff
@@ -0,0 +1,256 @@
+            md5sum = None
+            pass
+        return md5sum
+
+    try:
+        upload_resource = None
+        with open(resource['input_metadata_file']) as json_file:
+            data = json.load(json_file)
+        if disable_check:
+            # empty or non existing s3 folder
+            # skip individual s3 file check
+            upload_resource = resource
+            upload_resource['mime_type'] = data['mime_type']
+        else:
+            session = boto3.session.Session()
+            s3_client = session.client(
+                's3',
+                aws_access_key_id=aws_key,
+                aws_secret_access_key=aws_secret)
+            if data['s3_md5'] != s3_md5sum(s3_client, bucket, resource['output_s3']):
+                upload_resource = resource
+                upload_resource['mime_type'] = data['mime_type']
+        return upload_resource
+    except FileNotFoundError as e:
+        print('Error: No metadata json found!')
+        raise(e)
+
+
+def upload_s3(aws_key, aws_secret, filename, bucket, key, content_type):
+    """ upload s3 process for ThreadPoolExecutor """
+    # use session for multithreading according to
+    # https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html?highlight=multithreading#multithreading-multiprocessing
+    session = boto3.session.Session()
+    s3_client = session.client(
+        's3', aws_access_key_id=aws_key, aws_secret_access_key=aws_secret)
+    s3_client.upload_file(
+        Filename=filename,
+        Bucket=bucket,
+        Key=key,
+        ExtraArgs={
+            "ContentType": content_type
+        }
+    )
+    return key
+
+
+def upload(in_dir, bucket, bucket_folder):
+    """ upload resource and resource json to S3 """
+    aws_key = os.getenv('AWS_ACCESS_KEY_ID')
+    aws_secret = os.getenv('AWS_SECRET_ACCESS_KEY')
+
+    disable_deep_folder_check = is_s3_folder_empty(
+        aws_key=aws_key,
+        aws_secret=aws_secret,
+        bucket=bucket,
+        key=bucket_folder)
+
+    # environment variables mandatory for ThreadPoolExecutor
+    if aws_key is None or aws_secret is None:
+        raise EnvironmentError(
+            'Failed because AWS_ACCESS_KEY_ID and/or AWS_SECRET_ACCESS_KEY environment variable missing.')
+
+    sha1_filepattern = '?' * 40  # sha1 hexdigest as 40 characters
+
+    # get all resources on file system
+    all_resources = []
+    for sha1filename in Path(in_dir).glob(sha1_filepattern):
+        resource = {}
+        resource['bucket'] = bucket
+        resource['input_file'] = str(sha1filename)
+        resource['input_metadata_file'] = resource['input_file'] + '.json'
+        resource['output_s3'] = slash_join(
+            bucket_folder, os.path.basename(resource['input_file']))
+        resource['output_s3_metadata'] = slash_join(bucket_folder,
+                                                    os.path.basename(resource['input_file']) + '-unused.json')
```

Comment:
seems like Path/pathlib provides good alternatives to most of the operations you are doing here if that's something that interests you. if not, it doesn't bother me.

Reply:
Yes I'm oldschool and learned about Path/pathlib too late in the process of coding this module actually (I was unfamiliar with it basically). I simply sticked to os.path because I used it already a lot in the codebase and feared of hitting corner cases with rewriting it with pathlib. I will use pathlib for the future!